File: src/common/control.c
// 26 may 2015
#include "../ui.h"
#include "uipriv.h"
#include <stdio.h>

void uiControlDestroy(uiControl *c) {
	if (!c->Destroy)
		return;
	(*c->Destroy)(c);
}

uintptr_t uiControlHandle(uiControl *c) {
	if(!c->Handle)
		return NULL;
	return (*c->Handle)(c);
}

uiControl *uiControlParent(uiControl *c) {
	if (!c->Parent)
		return NULL;
	return (*c->Parent)(c);
}

void uiControlSetParent(uiControl *c, uiControl *parent) {
	if (!c->SetParent)
		return;
	(*c->SetParent)(c, parent);
}

int uiControlToplevel(uiControl *c) {
	if (!c->Toplevel)
		return 0;
	return (*c->Toplevel)(c);
}

int uiControlVisible(uiControl *c) {
	if (!c->Visible)
		return 0;
	return (*c->Visible)(c);
}

void uiControlShow(uiControl *c) {
	if (!c->Show)
		return;
	(*c->Show)(c);
}

void uiControlHide(uiControl *c) {
	if (!c->Hide)
		return;
	(*c->Hide)(c);
}

int uiControlEnabled(uiControl *c) {
	if (!c->Enabled)
		return 0;
	return (*c->Enabled)(c);
}

void uiControlEnable(uiControl *c) {
	if (!c->Enable)
		return;
	(*c->Enable)(c);
}

void uiControlDisable(uiControl *c) {
	if (!c->Disable)
		return;
	(*c->Disable)(c);
}

#define uiprivControlSignature 0x7569436F

uiControl *uiAllocControl(size_t size, uint32_t OSsig, uint32_t typesig, const char *typenamestr)
{
	uiControl *c;

	c = (uiControl *) uiprivAlloc(size, typenamestr);
	c->Signature = uiprivControlSignature;
	c->OSSignature = OSsig;
	c->TypeSignature = typesig;
	return c;
}

void uiFreeControl(uiControl *c) {
	if (uiControlParent(c) != NULL)
		uiprivUserBug("You cannot destroy a uiControl while it still has a parent. (control: %p)", c);
	uiprivFree(c);
}

void uiControlVerifySetParent(uiControl *c, uiControl *parent) {
	uiControl *curParent;

	if (uiControlToplevel(c))
		uiprivUserBug("You cannot give a toplevel uiControl a parent. (control: %p)", c);
	curParent = uiControlParent(c);
	if (parent != NULL && curParent != NULL)
		uiprivUserBug("You cannot give a uiControl a parent while it already has one. (control: %p; current parent: %p; new parent: %p)", c, curParent, parent);
	if (parent == NULL && curParent == NULL)
		uiprivImplBug("attempt to double unparent uiControl %p", c);
}

int uiControlEnabledToUser(uiControl *c) {
	while (c != NULL) {
		if (!uiControlEnabled(c))
			return 0;
		c = uiControlParent(c);
	}
	return 1;
}

File: src/control.rs
use {
    crate::raw::{
        uiAllocControl, uiControl, uiControlDestroy, uiControlDisable, uiControlEnable,
        uiControlEnabled, uiControlEnabledToUser, uiControlHandle, uiControlHide, uiControlParent,
        uiControlSetParent, uiControlShow, uiControlToplevel, uiControlVerifySetParent,
        uiControlVisible, uiFreeControl,
    },
    std::{
        ffi::{CString, NulError},
        ptr::null_mut,
    },
};

pub trait Control: AsRef<Self> {
    fn as_ptr(&self) -> *const uiControl {
        self.as_ptr_mut() as _
    }

    fn as_ptr_mut(&self) -> *mut uiControl;

    fn from_ptr(ptr: *mut uiControl) -> Self;

    /// Frees the memory associated with the control reference.
    ///
    /// # note
    /// This method is public only for writing custom controls.
    fn free(&self) {
        unsafe { uiFreeControl(self.as_ptr_mut()) }
    }

    /// Makes sure the control's parent can be set to `parent`.
    ///
    /// # arguments
    /// * `parent`: uiControl instance.
    ///
    /// # todo
    /// Make sure all controls have these
    ///
    /// # warning
    /// This will crash the application if `FALSE`.
    fn verify_set_parent<C, I>(&self, parent: C)
    where
        C: AsRef<I>,
        I: Control,
    {
        unsafe { uiControlVerifySetParent(self.as_ptr_mut(), parent.as_ref().as_ptr_mut()) }
    }

    /// Returns whether the control can be interacted with by the user.
    /// Checks if the control and all it's parents are enabled to make sure it can
    /// be interacted with by the user.
    ///
    /// # returns
    /// `true` if enabled, `FALSE` otherwise.
    ///
    /// # see
    /// - `uiControlEnabled`
    fn enabled_to_user(&self) -> bool {
        unsafe { uiControlEnabledToUser(self.as_ptr_mut()) != 0 }
    }

    /// Dispose and free all allocated resources.
    /// The platform specific APIs that actually destroy a control (and its children) are called.
    ///
    /// # note
    /// * Most of the time is needed to be used directly only on the top level windows.
    ///
    /// # todo
    /// Document ownership.
    fn destroy(&self) {
        unsafe { uiControlDestroy(self.as_ptr_mut()) }
    }

    /// Returns the control's OS-level handle.
    ///
    /// # returns
    /// * OS-level handle.
    fn handle(&self) -> usize {
        unsafe { uiControlHandle(self.as_ptr_mut()) }
    }

    /// Returns the parent control.
    ///
    /// # returns
    /// * The parent control, `NULL` if detached.
    fn parent<O>(&self) -> Option<O>
    where
        O: Control,
    {
        let ptr = unsafe { uiControlParent(self.as_ptr_mut()) };
        if ptr.is_null() {
            None
        } else {
            Some(O::from_ptr(ptr))
        }
    }

    /// Sets the control's parent.
    ///
    /// # arguments
    /// * `parent`: The parent control, `NULL` to detach.
    ///
    /// # todo
    /// Document ownership.
    fn set_parent<P, I>(&self, parent: Option<P>)
    where
        P: AsRef<I>,
        I: Control,
    {
        let parent = match parent {
            None => null_mut(),
            Some(p) => p.as_ref().as_ptr_mut(),
        };

        unsafe { uiControlSetParent(self.as_ptr_mut(), parent) }
    }

    /// Returns whether the control is a top level control.
    ///
    /// # returns
    /// * `true` if top level control, `FALSE` otherwise.
    fn toplevel(&self) -> bool {
        unsafe { uiControlToplevel(self.as_ptr_mut()) != 0 }
    }

    /// Returns whether the control is visible.
    ///
    /// # returns
    /// * `true` if visible, `FALSE` otherwise.
    fn visible(&self) -> bool {
        unsafe { uiControlVisible(self.as_ptr_mut()) != 0 }
    }

    /// Shows the control.
    fn show(&self) {
        unsafe { uiControlShow(self.as_ptr_mut()) }
    }

    /// Hides the control.
    ///
    /// # note
    /// * Hidden controls do not take up space within the layout.
    fn hide(&self) {
        unsafe { uiControlHide(self.as_ptr_mut()) }
    }

    /// Returns whether the control is enabled.
    /// Defaults to `true`.
    ///
    /// # see
    /// - `uiControlEnabledToUser`
    fn enabled(&self) -> bool {
        unsafe { uiControlEnabled(self.as_ptr_mut()) != 0 }
    }

    /// Enables the control.
    fn enable(&self) {
        unsafe { uiControlEnable(self.as_ptr_mut()) }
    }

    /// Disables the control.
    fn disable(&self) {
        unsafe { uiControlDisable(self.as_ptr_mut()) }
    }

    /// Allocates a uiControl.
    /// Helper to allocate new controls.
    ///
    /// # todo
    /// Document parameters
    fn alloc(os_sig: u32, type_sig: u32, type_name_str: &str) -> Result<Self, NulError>
    where
        Self: Sized,
    {
        let type_name_str = CString::new(type_name_str)?;
        let ptr = unsafe {
            uiAllocControl(
                size_of::<uiControl>(),
                os_sig,
                type_sig,
                type_name_str.as_ptr(),
            )
        };
        Ok(Self::from_ptr(ptr))
    }
}
